### PR TITLE
ENH: Commit history of files using GitRepo

### PR DESCRIPTION
**Pull request by debanjum**
Returns generator containing commit objects in reverse chronological order.
Useful for finding last commit date of files in dataset for web-interface metadata generator(`ls_json`) etc

Currently this is too slow for actually using on the datasets.datalad itself.
Being able to emulate `git rev-list` functionality through GitPython would be ideal.
Something like `git rev-list -1 HEAD --timestamp -- <filename>` for getting last commit timestamp

refer: [yarik's comment](https://github.com/datalad/datalad/pull/541/files/99d6483ec08afbc984bdbc87d8fc10d6a7df7495#r67255549)

### CODE REVIEW

File: datalad/support/gitrepo.py [hunk 1]
```diff
@@ -846,6 +845,23 @@ def get_file_content(self, file_, branch='HEAD'):
...
+          list of files, only commits with queried files are considered
+        branch: str
+          Name of the branch to query. Default: HEAD.
+
+        Returns
+        -------
+        [iterator]
+        yielding Commit items generator from branch history associated with files
+        """
+        return gitpy.objects.commit.Commit.iter_items(self.repo, branch, paths=files)
```
Comment (yarikoptic) [inline datalad/support/gitrepo.py]: Btw does this include merge commits as well? Just wondering

**Comment by bpoldrack** [comment]
In general I'd prefer to not expose GitPython objects outside `GitRepo`. Whoever deals with `GitRepo` shouldn't need to know about data structures used by GitPython, I think. Furthermore, I can imagine we drop GitPython at some point. So, could you just return the SHAs instead of `Commit`-objects?

**Comment by yarikoptic** [comment]
yeah, theoretically probably indeed better not to expose **but** we do already expose `.repo` of it anyways  :-/ and since we also have `Dataset.repo`  IIRC it is all confusing just to add an insult ;-)  may be at some point we could refactor it to `._git_repo` and see where we leak/use it.
having said that 
- as for history of the files  I don't think that SHAs would be sufficient for its initial intended purpose (discover last modification times)
- btw we do have also `GitRepo.get_branch_commits` which by default would return entire GitPython commit object I think but you could say `value="hexsha"` which would then return only the hexsha for the commit.  May be that function could be tuned up with `files=None` argument to perform what @debanjum  is trying to do here

May be eventually we would be doomed to  create our own presentation of a Commit object...

**Comment by debanjum** [comment]
Yeah I want the commit dates, so something more than the just the SHA would be needed. 

The GitRepo object is just a wrapper around Git's commit info so the underlying information it represents isn't GitRepo dependent per se. If and when we substitue GitRepo, the commit object it returns should still ideally contain the same set of information(as its ~= to Git's commit info).

And we are actually wrapping GitRepo with our `gitrepo.py`. So all we'll need to do in the function if we've to replace GitRepo is just call whatever new thing we're using from inside the `get_file_history` function and return a similar commit object.

File: datalad/support/gitrepo.py (hunk 1 above)
Comment (debanjum) [inline datalad/support/gitrepo.py]: No it doesn't seem to include merge commits. The commit on which the file was actually edited and commited is the ones in which are listed

**Comment by mih** [comment]
I would also prefer to not deepen our GitPython infiltration further than what we have now.

**Comment by yarikoptic** [comment]
@debanjum -- now we have a `get_committed_date` so it might make sense to 
- rename yours (which reveals GitPython) into protected (with _ prefix)
- provide `get_committed_dates` which would generate a dictionary with times per each requested file (or all files in the tree if no explicit filenames were requested), so it would all be inline and we wouldn't expose GitPython

**Comment by yarikoptic** [comment]
FTR: this function later to be used while generating tarballs and web views

**Comment by bpoldrack** [comment]
Anything new on this one? WIP label was added, but nothing happened since ...

**Comment by yarikoptic** [comment]
superseded by #1900

**Pull request closed without merging**